DataFormats/Portable/interface/PortableHostCollection.h

```diff
@@ -51,34 +125,106 @@ class PortableHostCollection {
+  ConstView<Idx> const& const_view() const {
+    return get<Idx>().view_;
+  }
+
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  View<Idx>& operator*() {
+    return get<Idx>().view_;
+  }
 
-  View& operator*() { return view_; }
-  ConstView const& operator*() const { return view_; }
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  ConstView<Idx> const& operator*() const {
+    return get<Idx>().view_;
+  }
+
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  View<Idx>* operator->() {
+    return &get<Idx>().view_;
+  }
 
-  View* operator->() { return &view_; }
-  ConstView const* operator->() const { return &view_; }
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  ConstView<Idx> const* operator->() const {
+    return &get<Idx>().view_;
+  }
+
+  // access the View by type
+  template <typename T>
+  typename T::View& view() {
+    return get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::ConstView const& view() const {
+    return get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::ConstView const& const_view() const {
+    return get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::View& operator*() {
+    return get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::ConstView const& operator*() const {
+    return get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::View* operator->() {
+    return &get<T>().view_;
+  }
+
+  template <typename T>
+  typename T::ConstView const* operator->() const {
+    return &get<T>().view_;
+  }
 
   // access the Buffer
   Buffer buffer() { return *buffer_; }
   ConstBuffer buffer() const { return *buffer_; }
   ConstBuffer const_buffer() const { return *buffer_; }
 
+  // Extract the sizes array
+  SizesArray sizes() const {
+    SizesArray ret;
+    constexpr_for<0, members_>([&](auto i) { ret[i] = get<i>().layout_.metadata().size(); });
+    return ret;
+  }
+
   // part of the ROOT read streamer
-  static void ROOTReadStreamer(PortableHostCollection* newObj, Layout const& layout) {
+  static void ROOTReadStreamer(PortableHostCollection* newObj, Implementation const& impl) {
     newObj->~PortableHostCollection();
     // use the global "host" object returned by cms::alpakatools::host()
-    new (newObj) PortableHostCollection(layout.metadata().size(), cms::alpakatools::host());
-    newObj->layout_.ROOTReadStreamer(layout);
+    std::array<int32_t, members_> sizes;
+    constexpr_for<0, members_>(
+        [&sizes, &impl](auto i) { sizes[i] = static_cast<Leaf<i> const&>(impl).layout_.metadata().size(); });
+    new (newObj) PortableHostCollection(sizes, cms::alpakatools::host());
+    constexpr_for<0, members_>([&sizes, &newObj, &impl](auto i) {
+      static_cast<Leaf<i>&>(newObj->impl_).layout_.ROOTReadStreamer(static_cast<Leaf<i> const&>(impl).layout_);
+    });
```

Comment:
@ericcano `sizes` does not seem to be used here; do we need it ?

Reply:
Indeed, it is not needed for the second `constexpr_for` loop on line 220.